internal/pkg/api/handleUpload.go

```diff
@@ -0,0 +1,381 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package api
+
+import (
+	"context"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+	"github.com/elastic/fleet-server/v7/internal/pkg/cache"
+	"github.com/elastic/fleet-server/v7/internal/pkg/config"
+	"github.com/elastic/fleet-server/v7/internal/pkg/dl"
+	"github.com/elastic/fleet-server/v7/internal/pkg/limit"
+	"github.com/elastic/fleet-server/v7/internal/pkg/logger"
+	"github.com/elastic/fleet-server/v7/internal/pkg/model"
+	"github.com/elastic/fleet-server/v7/internal/pkg/upload"
+	"github.com/elastic/fleet-server/v7/internal/pkg/upload/cbor"
+	"github.com/elastic/go-elasticsearch/v7"
+	"github.com/julienschmidt/httprouter"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+// the only valid values of upload status according to storage spec
+type UploadStatus string
+
+const (
+	UploadAwaiting UploadStatus = "AWAITING_UPLOAD"
+	UploadProgress UploadStatus = "UPLOADING"
+	UploadDone     UploadStatus = "READY"
+	UploadFail     UploadStatus = "UPLOAD_ERROR"
+	UploadDel      UploadStatus = "DELETED"
+)
+
+const (
+	// TODO: move to a config
+	maxParallelUploadOperations = 3
+	maxParallelChunks           = 4
+	maxFileSize                 = 104857600 // 100 MiB
+
+)
```

Comment:
I think server limits would be a good place to put this

Reply:
+1 to making these configurable, when this is released you will want a way to change them without having to wait for the next stack release.

Reply:
configurable or not "maxParallelUploadOperations = 3" is way too low.  there can be up to 125k endpoints attached to one fleet server.  We have to be able to handle much higher rates.

Reply:
> We have to be able to handle much higher rates

I think for the first release, we are specifically not looking to handle high throughput. When we have better purpose-built file storage solutions for this to flow into (native ES file storage, or being able to use an object store directly), we will certainly look to crank this up.

The UI is also built around this being a manual single-action tool. This is not designed to "fetch the same file from 50 endpoints". The user is fetching from one machine at a time. 

the client implementations were designed to buffer files for as long as desired, backoff, and retry if hitting limits

Reply:
;) If you build it, they will come.